fastestimator/trace/adapt/terminate_on_nan.py

```diff
@@ -0,0 +1,90 @@
+from fastestimator.trace.trace import Trace
+from fastestimator.util.data import Data
+from fastestimator.util.traceability_util import traceable
+from fastestimator.util.util import to_set
+
+
+@traceable()
+class TerminateOnNaN(Trace):
+    """End Training if a NaN value is detected. By default (inputs=None) it will monitor all loss values at the end
+    of each batch. If one or more inputs are specified, it will only monitor those values. Inputs may be loss keys
+    and/or the keys corresponding to the outputs of other traces (ex. accuracy).
+
+    Args:
+        monitor_names: key(s) to monitor for NaN values. If None, all loss values will be monitored. "*" will monitor
+            all trace output keys and losses.
+        mode: What mode(s) to execute this Trace in. For example, "train", "eval", "test", or "infer". To execute
+            regardless of mode, pass None. To execute in all modes except for a particular one, you can pass an argument
+            like "!infer" or "!train".
+    """
+    def __init__(self, monitor_names: Union[None, str, Set[str]] = None, mode: Union[None, str,
+                                                                                     Set[str]] = None) -> None:
+        self.monitored_keys = monitor_names if monitor_names is None else to_set(monitor_names)
+        super().__init__(inputs=self.monitored_keys, mode=mode)
+        self.all_loss_keys = {}
+        self.monitored_loss_keys = {}
+        self.monitored_trace_keys = {}
+        self.in_list = True
+
+    def on_epoch_begin(self, data: Data) -> None:
+        self.all_loss_keys = self.system.network.get_loss_keys()
+        if self.monitored_keys is None:
+            self.monitored_loss_keys = self.all_loss_keys
+        elif "*" in self.monitored_keys:
+            self.monitored_loss_keys = self.all_loss_keys
+            self.monitored_trace_keys = set()
+
+            for trace in get_current_items(self.system.traces, run_modes=self.system.mode, epoch=self.system.epoch_idx):
+                self.monitored_trace_keys.update(trace.outputs)
+        else:
+            self.monitored_loss_keys = self.monitored_keys & self.all_loss_keys
+            self.monitored_trace_keys = self.monitored_keys - self.monitored_loss_keys
+
+    def on_batch_end(self, data: Data) -> None:
+        for key in self.monitored_loss_keys:
+            if key in data:
+                if self._check_nan(data[key]):
+                    self.system.stop_training = True
+                    print("FastEstimator-TerminateOnNaN: NaN Detected in Loss: {}".format(key))
+        for key in self.monitored_trace_keys:
+            if key in data:
+                if self._check_nan(data[key]):
+                    self.system.stop_training = True
+                    print("FastEstimator-TerminateOnNaN: NaN Detected in: {}".format(key))
+
+    def on_epoch_end(self, data: Data) -> None:
+        for key in self.monitored_trace_keys:
```

Comment:
There might be NaN in evaluation loss. why only trace_keys here?

Reply:
The ideas was that on_batch_end method will handle nan losses in eval mode (similar to train data) but some traces outputs could only be available on epoch_end, so checking for trace keys on_epoch_end. Hence, the case. However, we can check that at epoch end too and in that case I can combine both losses and trace keys as well.